src/forms/FormControl.scss

```diff
@@ -104,166 +138,197 @@
-      background-color: transparent;
-      border: solid var(--primer-borderWidth-thin, 1px) transparent;
-      border-radius: var(--primer-borderRadius-medium, 6px);
-      transition: 0.2s cubic-bezier(0.3, 0, 0.5, 1);
-      transition-property: color, background-color, border-color;
-      grid-row: 1;
-      place-content: center;
-      place-self: center end;
-      grid-column: 2 / 3;
-
-      &:hover {
-        background-color: var(--color-btn-hover-bg);
-        border: solid var(--primer-borderWidth-thin, 1px) var(--color-btn-hover-bg);
+  // size modifications can be refactored with :has() - FormControl-fieldWrap:has(.FormControl--large)
+  // sizes
+  &.FormControl-fieldWrap--input {
+    &.FormControl--small {
+      .FormControl--input-leadingVisualWrap {
+        top: calc(var(--primer-control-medium-paddingInline-condensed, 8px) - 0.125rem); /* 6px */
+        left: calc(var(--primer-control-medium-paddingInline-condensed, 8px) - 0.125rem); /* 6px */
       }
 
-      &[disabled] {
-        color: var(--color-primer-fg-disabled);
-        pointer-events: none;
+      &.FormControl-fieldWrap--input-trailingAction {
+        .FormControl--input.FormControl--small {
+          padding-inline-end:
+            calc(
+              var(----primer-control-small-paddingInline-condensed, 8px) + var(--base-size-16, 16px) +
+              var(--primer-control-small-gap, 8px) - var(--primer-borderWidth-thin, 1px)
+            ); /* 27px */
+        }
       }
-    }
-
-    // sizes
-
-    // these selectors can be refactored to use :has()
-    &.FormControl--small {
-      height: var(--primer-control-small-size, 28px);
-      font-size: var(--primer-text-body-size-small, 12px);
 
       .FormControl--input-trailingAction {
         width: calc(var(--primer-control-small-size, 28px) - var(--base-size-8, 8px));
         height: calc(var(--primer-control-small-size, 28px) - var(--base-size-8, 8px));
+
+        &::before {
+          top: calc((var(--primer-control-xsmall-size) - var(--base-size-16)) / 4); /* 2px */
+        }
       }
     }
 
-    &.FormControl--medium {
-      height: var(--primer-control-medium-size, 32px);
+    &.FormControl--large {
+      .FormControl--input-leadingVisualWrap {
+        top: var(--primer-control-medium-paddingInline-normal, 12px);
+        left: var(--primer-control-medium-paddingInline-normal, 12px);
+      }
 
-      .FormControl--input-trailingAction {
-        width: calc(var(--primer-control-medium-size, 32px) - var(--base-size-8, 8px));
-        height: calc(var(--primer-control-medium-size, 32px) - var(--base-size-8, 8px));
+      &.FormControl-fieldWrap--input-leadingVisual {
+        .FormControl--input.FormControl--large {
+          padding-inline-start:
+            calc(
+              var(--primer-control-large-paddingInline-normal, 12px) + var(--base-size-16, 16px) +
+              var(--primer-control-large-gap, 8px) - var(--primer-borderWidth-thin, 1px)
+            ); /* 31px */
```

Comment:
Non-blocking nitpick: instead of commenting with the calculated value, it might be nice to add a comment explaining the equation.